packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -0,0 +1,219 @@
+  const parser = isTS ? makeParser('tsx') : makeParser('jsx');
+  // If this errors, it will be caught in the calling function, where we know more information and can construct a
+  // better warning message
+  return jscs(code, { parser });
+}
+
+/**
+ * Find all nodes which represent Identifiers with the given name
+ *
+ * @param ast The code, in AST form
+ * @param name The Identifier name to search for
+ * @returns A collection of NodePaths pointing to any nodes which were found
+ */
+export function findIdentifiers(ast: AST, name: string): AST<IdentifierNode> {
+  const identifierFilter = function (path: ASTPath<IdentifierNode>): boolean {
+    // Check that what we have is indeed an Identifier, and that the name matches
+    //
+    // Note: If we were being super precise about this, we'd also check the context in which the identifier is being
+    // used, because there are some cases where we actually don't want to be renaming things (if the identifier is being
+    // used to name a class property, for example). But the chances that someone is going to have a class property in a
+    // nextjs page file with the same name as one of the canonical functions are slim to none, so for simplicity we can
+    // stop filtering here. If this ever becomes a problem, more precise filter checks can be found in a comment at the
+    // bottom of this file.
+    return path.node.name === name;
+  };
+
+  return ast.find(Identifier).filter(identifierFilter);
+}
+
+/**
+ * Find all nodes which are declarations of variables with the given name
+ *
+ * @param ast The code, in AST form
+ * @param name The variable name to search for
+ * @returns A collection of NodePaths pointing to any nodes which were found
+ */
+export function findDeclarations(ast: AST, name: string): AST<VariableDeclarationNode> {
+  // Check for a structure of the form
+  //
+  //     node: VariableDeclaration
+  //      \
+  //       declarations: VariableDeclarator[]
+  //        \
+  //         0 : VariableDeclarator
+  //          \
+  //           id: Identifier
+  //            \
+  //             name: string
+  //
+  // where `name` matches the given name.
+  const declarationFilter = function (path: ASTPath<VariableDeclarationNode>): boolean {
+    return (
+      path.node.declarations.length === 1 &&
+      VariableDeclarator.check(path.node.declarations[0]) &&
+      Identifier.check(path.node.declarations[0].id) &&
+      path.node.declarations[0].id.name === name
+    );
+  };
+
+  return ast.find(VariableDeclaration).filter(declarationFilter);
+}
+
+/**
+ * Find all nodes which are exports of variables with the given name
+ *
+ * @param ast The code, in AST form
+ * @param name The variable name to search for
+ * @returns A collection of NodePaths pointing to any nodes which were found
+ */
+export function findExports(ast: AST, name: string): AST<ExportSpecifierNode> {
+  const exportFilter = function (path: ASTPath<ExportSpecifierNode>): boolean {
```

Comment:
nit: Why isn't this just an arrow function?

IMO if we aren't using the functionality that comes from a `function` declaration (bindings to `this`/`arguments`, using call, apply and bind methods, using `yield`) we should stick with an arrow function. This means if/when a `function` keyword is used anonymously - it'll indicate strongly to a user that something is interesting is going on.

Reply:
Huh. That’s a distinction I've never heard anyone make before. There are definitely times when I think one or the other is better, but that's almost always just a matter of readability. (There's literally almost almost nothing harder to parse than a typed higher-order arrow function whose return value is also written as an arrow function, especially if you have a few of them nested - it's just an utter sea of parentheses and colons and arrows. Most situations aren't to that extreme, of course, but the need to easily be able to tell what's what remains.) My rule of thumb is generally to use arrow functions for fully anonymous callbacks and very short function expressions (because in both cases it's pretty easy to tell at a glance that what you're looking at is a function) but to use the `function` keyword otherwise (especially if I'm already inside of another function) because it makes it easier to visually separate the function's code into its own semantic chunk.

Here I marginally prefer `functions`, for the reasons above, but don't think it's all that bad either way. So if it really bugs you, I'm happy to change it. What's your LOGAF here?

Reply:
This is more low than anything for me, I’ll let you make the call!